### ISSUE TITLE
Dialyzer specdiffs triggers missing_range warning on protocol typespecs

### ISSUE THREAD

**Issue by rodrigues** [elixir-lang/elixir#9685]
### Environment

* Elixir & Erlang/OTP versions (elixir --version):  1.9.4 / 22.1.8
* Operating system: macOS Catalina

### Current behavior

When using `specdiffs` dialyzer flag, it indicates protocol `__impl__/1` typespecs are wrong.

The code is below, I've created a repo so you can also reproduce the error: https://github.com/rodrigues/protocol_missing_range

```elixir
defprotocol Protocolo do
  def foo(thing)
end

defimpl Protocolo, for: Map do
  def foo(%{a: a}) do
    {a, 1}
  end
end

defimpl Protocolo, for: Tuple do
  def foo({a}) do
    {a, 1}
  end
end
```

```shell
➜ mix dialyzer
Compiling 1 file (.ex)
Generated protocol_missing_range app
Finding suitable PLTs
Checking PLT...
[:compiler, :elixir, :kernel, :logger, :stdlib]
PLT is up to date!
No :ignore_warnings opt specified in mix.exs and default does not exist.

Starting Dialyzer
[
  check_plt: false,
  init_plt: '/Users/rodrigues/code/protocol_missing_range/_build/dev/dialyxir_erlang-22.1.8_elixir-1.9.4_deps-dev.plt',
  files: ['/Users/rodrigues/code/protocol_missing_range/_build/dev/lib/protocol_missing_range/ebin/Elixir.Protocolo.Map.beam',
   '/Users/rodrigues/code/protocol_missing_range/_build/dev/lib/protocol_missing_range/ebin/Elixir.Protocolo.Tuple.beam',
   '/Users/rodrigues/code/protocol_missing_range/_build/dev/lib/protocol_missing_range/consolidated/Elixir.Protocolo.beam'],
  warnings: [:specdiffs, :unknown]
]
Total errors: 2, Skipped: 0, Unnecessary Skips: 0
done in 0m1.21s
lib/protocol_missing_range.ex:5:missing_range
The type specification is missing types returned by function.

Function:
Protocolo.Map.__impl__/1

Type specification return types:
Map

Extra types in success typing:
Protocolo | Protocolo.Map

________________________________________________________________________________
lib/protocol_missing_range.ex:11:missing_range
The type specification is missing types returned by function.

Function:
Protocolo.Tuple.__impl__/1

Type specification return types:
Tuple

Extra types in success typing:
Protocolo | Protocolo.Tuple

________________________________________________________________________________
done (warnings were emitted)
Halting VM with exit status 2
```

### Expected behavior

Correct typespecs and no dialyzer errors 😄

It looks more like an elixir issue, [probably here](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/protocol.ex#L832) or [here](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/protocol.ex#L900)?

Thanks, and happy new year! 🌈

**Comment by josevalim** [comment]
I don't believe we should fix this one. That's because the fix is to make the type specification *less* specific. We would have to rewrite:

```elixir
        @spec __impl__(:for) :: unquote(for)
        @spec __impl__(:target) :: __MODULE__
        @spec __impl__(:protocol) :: unquote(protocol)
```

to:

```elixir
        @spec __impl__(:for | :target | :protocol) :: unquote(for) | __MODULE__ | unquote(protocol)
```

FWIW, both "overspecs" and "specdiffs" have a warn in the official docs:

> The following options are also available, but their use is not recommended (they are mostly for Dialyzer developers and internal debugging):

That's because by using overspecs, you will be bound by how well Dialyzer can infer the types, even if you know better.

In any case, thanks for the report and happy new year too! 🎉

**Comment by rodrigues** [comment]
@josevalim thanks for the explanation!

That's just unfortunate, as it's only with `specdiffs` flag that I can get dialyzer to error on this:

```elixir
  @spec errorable_fun(term()) :: {:ok, term()}
  def errorable_fun(something) do
    if something do
      {:ok, something}
    else
      {:error, :something_else}
    end
  end
```

```
lib/dialyzer_examples.ex:38:missing_range
The type specification is missing types returned by function.

Function:
DialyzerExamples.errorable_fun/1

Type specification return types:
({:ok, _})

Extra types in success typing:
({:error, :something_else})
```

I was ignoring `~r/\:contract_subtype/` kind of errors that `specdiffs` might also trigger, since these are just way too much dialyzing, and unpractical for a dynamic language (maybe just for debugging), but so far the only place I've got a `missing_range` error, in a couple dozen repos, was in `__impl__/1` protocol typespecs; but now I get your point.

However there is still hope hehe, I'll try to reproduce this behaviour in an erlang repo, and maybe that could be improved in dialyzer side. Thanks a lot 👍

**Comment by rodrigues** [comment]
Happy ending after all: I could reproduce the error in erlang, but it got fixed in 22.2.1 🤘🍺

```erlang
-module(test).
-export([foo/1]).

-spec foo(integer()) -> a
       ; (binary()) -> b.
foo(A) when is_integer(A) ->
    a;
foo(B) when is_binary(B) ->
    b.
```

```shell
# OTP 22.1.8
➜ dialyzer -Wspecdiffs test.erl
  Checking whether the PLT /Users/rodrigues/.dialyzer_plt is up-to-date... yes
  Proceeding with analysis...
test.erl:4: The success typing for test:foo/1 implies that the function might also return
          'b' but the specification return is
          'a'
 done in 0m0.37s
done (warnings were emitted)
```

```shell
# OTP 22.2.1
➜ dialyzer -Wspecdiffs test.erl
  Checking whether the PLT /Users/rodrigues/.dialyzer_plt is up-to-date... yes
  Proceeding with analysis... done in 0m0.10s
done (passed successfully)
```
